File: 88.js
/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
const merge = (nums1, m, nums2, n) => {
	while (n) {
		if (nums1[m - 1] > nums2[n - 1]) {
			nums1[m + n - 1] = nums1[m - 1];
			m--;
		} else {
			nums1[m + n - 1] = nums2[n - 1];
			n--;
		}
	}
};

console.log(merge([1, 2, 3, 0, 0, 0], 3, [2, 5, 6], 3)); // [1, 2, 2, 3, 5, 6]
console.log(merge([1], 1, [], 0)); // [1]
console.log(merge([0], 0, [1], 1)); // [1]

File: 88.py
'''
Time: O(n+m)
Space: O(1)
'''
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        result = []
        p1 = 0
        p2 = 0

        while p1 < m and p2 < n:
            if nums1[p1] < nums2[p2]:
                result.append(nums1[p1])
                p1 += 1
            else:
                result.append(nums2[p2])
                p2 += 1

        for k in range(p1, m):
            result.append(nums1[k])
        for k in range(p2, n):
            result.append(nums2[k])

        for i in range(len(result)):
            nums1[i] = result[i]

        return

'''
Time: O(n+m)
Space: O(1)
'''
class Solution2:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        p1 = m - 1
        p2 = n - 1
        result = m + n - 1

        while p1 >= 0 and p2 >= 0:
            if nums1[p1] > nums2[p2]:
                nums1[result] = nums1[p1]
                p1 -= 1
            else:
                nums1[result] = nums2[p2]
                p2 -= 1

            result -= 1

        while p2 >= 0:
            nums1[result] = nums2[p2]
            p2 -= 1
            result -= 1

        return